internal/peer/node/throttle_grpc.go

```diff
@@ -0,0 +1,65 @@
+/*
+Copyright IBM Corp. All Rights Reserved.
+
+SPDX-License-Identifier: Apache-2.0
+*/
+
+package node
+
+import (
+	"context"
+
+	"github.com/hyperledger/fabric/common/semaphore"
+	"github.com/hyperledger/fabric/core/peer"
+	"github.com/pkg/errors"
+	"google.golang.org/grpc"
+)
+
+func initGrpcSemaphores(config *peer.Config) map[string]semaphore.Semaphore {
+	semaphores := make(map[string]semaphore.Semaphore)
+	endorserConcurrency := config.LimitsConcurrencyEndorserService
+	deliverConcurrency := config.LimitsConcurrencyDeliverService
+
+	if endorserConcurrency != 0 {
+		semaphores["/protos.Endorser/ProcessProposal"] = semaphore.New(endorserConcurrency)
+	}
+
+	if deliverConcurrency != 0 {
+		deliverSemaphore := semaphore.New(deliverConcurrency)
+		semaphores["/protos.Deliver/Deliver"] = deliverSemaphore
+		semaphores["/protos.Deliver/DeliverFiltered"] = deliverSemaphore
+		semaphores["/protos.Deliver/DeliverWithPrivateData"] = deliverSemaphore
+	}
+
+	return semaphores
+}
+
+func unaryThrottleInterceptor(semaphores map[string]semaphore.Semaphore) grpc.UnaryServerInterceptor {
+	return func(ctx context.Context, req interface{}, info *grpc.UnaryServerInfo, handler grpc.UnaryHandler) (interface{}, error) {
+		sema, ok := semaphores[info.FullMethod]
+		if !ok {
+			return handler(ctx, req)
+		}
+		if err := sema.AcquireNow(ctx); err != nil {
+			logger.Errorf("failed to process grpc request for %s: %s", info.FullMethod, err)
+			return nil, errors.WithMessagef(err, "failed to process grpc request for %s", info.FullMethod)
```

Comment:
@cendhu In this case, AcquireNow may return an error due to context canceled. Therefore, we are not sure if the error is too many request unless we introduce a new error type.

Reply:
Rethinking about it, decide to remove context from AcquireNow so that it only returns error when the buffer is full. Context cancellation will be caught at grpc layer anyway. Error message is updated accordingly.